q2cli/core/usage.py

```diff
@@ -0,0 +1,114 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2020, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from qiime2.core.type.primitive import Bool
+
+import qiime2.sdk.usage as usage
+from qiime2.sdk.util import is_metadata_type
+
+from q2cli.util import to_cli_name
+from q2cli.util import to_snake_case
+
+
+class CLIUsage(usage.Usage):
+    def __init__(self):
+        super().__init__()
+        self._recorder = []
+        self._init_data_refs = dict()
+        self._metadata_refs = dict()
+        self._col_refs = dict()
+
+    def _init_data_(self, ref, factory):
+        self._init_data_refs[ref] = factory
+        # Don't need to compute anything, so just pass along the ref
+        return ref
+
+    def _merge_metadata_(self, ref, records):
+        merge_target = ref
+        for record in records:
+            self._metadata_refs[record.ref] = record.result
+        return merge_target
+
+    def _get_metadata_column_(self, ref, record, column_name):
+        self._metadata_refs[record.ref] = record.result
+        self._col_refs[record.ref] = column_name
+        return ref
+
+    def _comment_(self, text: str):
+        self._recorder.append('# %s' % (text,))
+
+    def _action_(
+        self, action: usage.UsageAction, input_opts: dict, output_opts: dict
+    ):
+        action_f, action_sig = action.get_action()
+        t = self._template_action(action_f, input_opts, output_opts)
+        self._recorder.append(t)
+        return output_opts
+
+    def _assert_has_line_matching_(self, ref, label, path, expression):
+        pass
+
+    def render(self):
+        return '\n'.join(self._recorder)
+
+    def get_example_data(self):
+        return {r: f() for r, f in self._init_data_refs.items()}
+
+    def _template_action(self, action_f, input_opts, outputs):
+        cmd = to_cli_name(f"qiime {action_f.plugin_id} {action_f.id}")
+        inputs = template_inputs(action_f, input_opts)
+        params = template_parameters(action_f, input_opts)
+        params += template_metadata(self._metadata_refs, self._col_refs)
```

Comment:
I think you might be missing some metadata cases, check the following:

https://github.com/qiime2/qiime2/blob/7061b9896708cf2f5457c76857adc64ec0d0e021/qiime2/core/testing/examples.py#L133-L142

Reply:
Yep, those are coming soon.